app/services/gps_report.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+class InvalidSupplierError < StandardError; end
+
+class GPSReport
+  include ActionView::Helpers::TextHelper
+
+  def initialize(date_range, supplier_name)
+    @supplier = Supplier.find_by(key: supplier_name)
+    raise InvalidSupplierError, "Invalid supplier: `#{supplier_name}`. Valid values are: `#{Supplier.all.map(&:key).join('`, `')}`." if @supplier.nil?
+
+    @from_date = date_range.first
+    @to_date = date_range.last
+  end
+
+  def generate
+    Rails.logger.info "Generating GPS Data Report for #{@supplier.key} from #{@from_date} to #{@to_date}."
+
+    if journeys.empty?
+      Rails.logger.info "No #{@supplier.key} journeys found for the specified period."
+      return {
+        failures: [],
+        move_count: 0,
+      }
+    end
+
+    Rails.logger.info "Found #{pluralize(moves.count, "#{@supplier.key} move")} with #{pluralize(journeys.count, 'journey')}."
+
+    {
+      failures: failures,
+      move_count: moves.count,
+    }
+  end
+
+private
+
+  def failures
+    return @failures unless @failures.nil?
+
+    @failures = moves.filter_map { |move|
+      { move: move, reason: 'no_journeys' } if move.journeys.empty?
+    }.concat(journey_failures)
+  end
+
+  def journey_failures
+    failures = journeys.filter_map do |journey|
+      gps_data = gps_data_map[journey.id]
+      next { move: journey.move, reason: 'no_gps_data' } if gps_data.blank?
+
+      intervals = gps_data.each_with_index.filter_map { |t, i| i < gps_data.count - 1 ? gps_data[i + 1] - t : nil }
+      { move: journey.move, reason: 'gps_data_gap' } if intervals.max > 60.seconds
+    end
+    failures.uniq { |f| f[:move] }
+  end
+
+  def gps_data_map
+    return @gps_data_map unless @gps_data_map.nil?
+
+    repair_table
+
+    Rails.logger.info "Fetching #{@supplier.key} GPS data points."
+
+    time_since = TimeSince.new
+    res = athena_query("select journey_id, tracking_timestamp from #{@supplier.key} where journey_id in ('#{journeys.map(&:id).join("', '")}') order by journey_id, tracking_timestamp;")
+
+    @gps_data_map = {}
+    res.result_set.rows[1..].each do |data_row|
+      journey_id, timestamp = *data_row.data.map(&:var_char_value)
+      @gps_data_map[journey_id] = [] if @gps_data_map[journey_id].nil?
+
+      @gps_data_map[journey_id] << Time.zone.parse(timestamp)
+    end
```

Comment:
```suggestion
    @gps_data_map = res.result_set.rows[1..].each_with_object do |data_row, data_map|
      journey_id, timestamp = *data_row.data.map(&:var_char_value)
      data_map[journey_id] = [] if data_map[journey_id].nil?

      data_map[journey_id] << Time.zone.parse(timestamp)
    end
```

I think this is slightly nicer as it avoids needing to assign the variable to an empty hash first.

Reply:
Wow, nice!